ta: pkcs11: Load persistent object attributes during db initialization

rvdgracht opened the pull request:
Persistent object attributes are now loaded during the first find
operation. This means that you're forced to do a find operation to make
the objects attributes available.

Right now entry_find_objects_init is broken in such that it tries to
access the obj->attributes for objects in the session->token->object_list
before setting the obj->attributes (by calling token_obj_matches_ref).
This results in a Panic.

This patch fixes both issues.

Signed-off-by: Robin van der Gracht <robin@protonic.nl>

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

etienne-lms commented:
This change makes the TA using memory to load all persistent objects.
I think it would better to load in memory only the active objects, preferably only the visible objects.

> This results in a Panic.

You're definitely right. Late review changes in the P-R made call ordering being change, and we missed that.
Yet, I think the implementation here is overkilling.
We could call `load_object_attribs()` from `entry_find_objects_init()` ([here](https://github.com/OP-TEE/optee_os/blob/dc99b202f3997248adbe7508698c051f6d677cd0/ta/pkcs11/src/object.c#L622) before attributes are accessed in the loops) upon `obj->attributes` being NULL.

What do you think?
@ruchi393, @vesajaaskelainen , may I ask your opinion.

Note : this P-R deserves the **Bug** tag.

jforissier commented:
> Note : this P-R deserves the **Bug** tag.

I don't mind much, but PRs usually do not remain open for long, except for complex new features, so I'm not sure the tags are useful on PRs. Issues are a bit different due to the numerous user questions.

rvdgracht commented:
Thank you for reviewing.

>  This change makes the TA using memory to load all persistent objects.

True. Isn't that also what should happen the first time `entry_find_objects_init()` is called?

> I think it would better to load in memory only the active objects, preferably only the visible objects.

Is there a way to determine this without loading the attributes first? Or do you want to conditionally release
them afterwards?

etienne-lms commented:
The current implementation tries to keep a open handle on TEE objects only when needed.
This because each open TEE objects in the TA means resources allocated in Core.
So `obj->attribs_hdl` should be closed once the object (in core secure storage) no more needs to be accessed.

etienne-lms commented:
Based on your change, i made this fixup commit: https://github.com/etienne-lms/optee_os/commit/49620e996ef4eff9cb1f3a4d7d37748635104a49.
See my proposal + all minor changes I made that you can use as review comment for your change.

ruchi393 commented:
> This change makes the TA using memory to load all persistent objects.
> I think it would better to load in memory only the active objects, preferably only the visible objects.
> 
> > This results in a Panic.
> 
> You're definitely right. Late review changes in the P-R made call ordering being change, and we missed that.
> Yet, I think the implementation here is overkilling.
> We could call `load_object_attribs()` from `entry_find_objects_init()` ([here](https://github.com/OP-TEE/optee_os/blob/dc99b202f3997248adbe7508698c051f6d677cd0/ta/pkcs11/src/object.c#L622) before attributes are accessed in the loops) upon `obj->attributes` being NULL.
> 
> What do you think?
> @ruchi393, @vesajaaskelainen , may I ask your opinion.
> 
> Note : this P-R deserves the **Bug** tag.

@etienne-lms , I reviewed the patch - etienne-lms@49620e9. The approach suggested by you looks good to me. Saves the extra attributes of persistent objects in memory when not needed.

vesajaaskelainen commented:
> This change makes the TA using memory to load all persistent objects.
> I think it would better to load in memory only the active objects, preferably only the visible objects.
> 
> > This results in a Panic.
> 
> You're definitely right. Late review changes in the P-R made call ordering being change, and we missed that.
> Yet, I think the implementation here is overkilling.
> We could call `load_object_attribs()` from `entry_find_objects_init()` ([here](https://github.com/OP-TEE/optee_os/blob/dc99b202f3997248adbe7508698c051f6d677cd0/ta/pkcs11/src/object.c#L622) before attributes are accessed in the loops) upon `obj->attributes` being NULL.
> 
> What do you think?
> @ruchi393, @vesajaaskelainen , may I ask your opinion.
> 
> Note : this P-R deserves the **Bug** tag.

I will only tackle the memory issue here for now.

We actually increased the heap size to 64k just in case there are more keys in used. We once hit the memory allocation barrier with the TA so decided that let's pick a bit larger number so that we do not hit it in normal use cases. What attributed to that hitting the barrier was that we tried to test it out by creating more keys than what we normally would have in the device. I enabled the memory debugging facilities and noticed that there were actually leftovers of decrypted indices of secure storage objects in the memory (still allocated). Don't know if this is already fixed or not in upstream code base.

Our logic was a bit that we do have 32 MiB allocation for TEE OS. So we can actually use the area.

But what is currently a problem is that the operations are actually quite slow. Eg. if you make multiple TLS connections it is rather slow to initiate the operations for those. Round trip from REE -> TEE -> TA -> RPC:TEE -> RPC: REE -> TEE -> TA -> actual operation has some kind of cost. So some kind of cache would be very nice to have.

While thinking about the cache is the what to cache. Should it be operation in TEE layer so that read-only copies can be cached? Should it be in TA or in both? Should it be in encrypted form or in decrypted form in memory. AND in what memory. What we are here trying to do here is protect the key material so there are lot of vectors to think.

rvdgracht commented:
`Based on your change, i made this fixup commit: etienne-lms@49620e9`

ACK. This is more efficiënt memory wise. I'll try to test this on my hardware tomorrow :)

`So some kind of cache would be very nice to have.`

@vesajaaskelainen Maybe we should create a new issue for this?

vesajaaskelainen commented:
> `So some kind of cache would be very nice to have.`
> 
> @vesajaaskelainen Maybe we should create a new issue for this?

Sure no need to address it in this ticket.

rvdgracht commented:
@etienne-lms I've squashed our patches and updated the commit message. I've tested this on my hardware and found no issues.
I've included both our s-o-b in the commit message. Let me know if thats ok with you.

etienne-lms commented:
> I've included both our s-o-b in the commit message. Let me know if thats ok with you.

You can discard my S-o-b tag from the change if you want. It's not really needed, my proposal was just a review comment (that I found easier to share through a patch rather than unitary github review comments).

That said, looks good to me (w/ or w/o my S-o-b) with a minor coding style comment.
(Note I can't give a R-b or A-b tag if the change includes my S-o-b tag :)

rvdgracht commented:
> You can discard my S-o-b tag from the change

I've fixed the code style issue Travis complained about and changed the S-o-b into a R-b.
I've pushed the changes to our [early-load-db-obj-attr](https://github.com/Protonic-Holland/optee_os/tree/early-load-db-obj-attr) branch but its not (yet?) showing on this page.

rvdgracht commented:
> You can discard my S-o-b tag from the change

I've fixed the code style issue Travis complained about and changed the S-o-b into a R-b.
I've pushed the changes to our [early-load-db-obj-attr](https://github.com/Protonic-Holland/optee_os/tree/early-load-db-obj-attr) branch but its not (yet?) showing on this page.

jforissier commented:
@etienne-lms let me know if you're OK with the R-b tag or if you'd rather have a A-b here.

ruchi393 commented:
Acked-by: Ruchika Gupta <ruchika.gupta@linaro.org>

etienne-lms commented:
@jforissier, my R-b applies here. Thanks.

rvdgracht commented:
> Acked-by: Ruchika Gupta ruchika.gupta@linaro.org

Added to the commit message

jforissier commented:
> > Acked-by: Ruchika Gupta [ruchika.gupta@linaro.org](mailto:ruchika.gupta@linaro.org)
> 
> Added to the commit message

Wrong format, missing `<>` it seems (see checkpatch error in Travis CI)

rvdgracht commented:
> Wrong format, missing <> it seems

Copy paste issue. Should be fixed now.
